File: ChatGPT(예진)/Cpp/미니RPG전투게임.cpp
#include <iostream>
#include <string>
#include <map>
#include <vector>
#include <random>

class Character {
public:
    std::string name;
    int level;
    int maxHealth;
    int health;
    int attack;
    int experience;
    std::map<std::string, int> items;
    std::vector<Character*> allies;

    Character(std::string name, int level) : name(name), level(level) {
        maxHealth = rand() % 51 + 50 + (10 * level);
        health = maxHealth;
        attack = rand() % 16 + 5 + (2 * level);
        experience = 0;
        items = { {"health_potion", 3}, {"mana_potion", 2} };
    }

    virtual void takeDamage(int damage) {
        health -= damage;
        if (health < 0) {
            health = 0;
        }
    }

    bool isAlive() {
        return health > 0;
    }

    void levelUp() {
        level++;
        maxHealth += 10;
        attack += 2;
        health = maxHealth;
        std::cout << name << " leveled up to level " << level << "!" << std::endl;
    }

    void gainExperience(int exp) {
        experience += exp;
        std::cout << name << " gained " << exp << " experience points!" << std::endl;
        if (experience >= level * 100) {
            levelUp();
        }
    }

    void useItem(std::string item) {
        if (items.find(item) != items.end() && items[item] > 0) {
            if (item == "health_potion") {
                int healthToRestore = std::min(30, maxHealth - health);
                health += healthToRestore;
                items[item]--;
                std::cout << name << " used a health potion and restored " << healthToRestore << " health!" << std::endl;
            }
            else if (item == "mana_potion") {
                items[item]--;
                std::cout << name << " used a mana potion!" << std::endl;
                // Add mana restoring logic here if needed
            }
        }
    }

    std::string status() {
        return name + "'s health: " + std::to_string(health) + "/" + std::to_string(maxHealth) +
            ", Level: " + std::to_string(level) + ", Experience: " + std::to_string(experience) +
            "/" + std::to_string(level * 100) + ", Items: {health_potion: " + std::to_string(items["health_potion"]) +
            ", mana_potion: " + std::to_string(items["mana_potion"]) + "}";
    }

    void joinAlly(Character* ally) {
        allies.push_back(ally);
        std::cout << ally->name << " joined " << name << "'s party!" << std::endl;
    }
};

class Player : public Character {
public:
    int gold;

    Player(std::string name, int level) : Character(name, level), gold(100) {}

    void takeDamage(int damage) {
        Character::takeDamage(damage);
        if (attack < 0) {
            attack = 0;
        }
    }
};

class Enemy : public Character {
public:
    std::string specialAbility;

    Enemy(std::string name, int level, std::string specialAbility) : Character(name, level), specialAbility(specialAbility) {}

    void takeDamage(int damage) {
        if (specialAbility == "block") {
            std::cout << name << " blocks the attack!" << std::endl;
        }
        else {
            Character::takeDamage(damage);
        }
    }

    void attackTarget(Character* target) {
        if (specialAbility == "stun") {
            std::cout << name << " uses stun ability!" << std::endl;
            target->takeDamage(attack * 2);
        }
        else {
            target->takeDamage(std::max(1, attack));
        }
    }
};

void randomEvent(Player* player, Enemy* enemy) {
    double eventChance = static_cast<double>(rand()) / RAND_MAX; // Generate a random number between 0 and 1
    if (eventChance < 0.2) { // 20% chance of an event occurring
        int event = rand() % 5 + 1;
        switch (event) {
        case 1:
            std::cout << "A friendly wizard appears and heals you." << std::endl;
            player->health = std::min(player->maxHealth, player->health + 20);
            break;
        case 2:
            std::cout << "You found a treasure chest! You gained 50 experience points." << std::endl;
            player->gainExperience(50);
            break;
        case 3:
            std::cout << "A curse befalls you, reducing your attack power!" << std::endl;
            if (player->attack >= 5) {
                player->attack -= 5;
            }
            else {
                player->attack = 0;
            }
            break;
        case 4:
            std::cout << "You stumbled upon a hidden passage and found a health potion!" << std::endl;
            player->useItem("health_potion");
            break;
        default:
            std::cout << "A mysterious mist surrounds you, confusing the enemy!" << std::endl;
            if (enemy->attack >= 5) {
                enemy->attack -= 5;
            }
            else {
                enemy->attack = 0;
            }
            break;
        }
    }
}

void battle(Player* player, Enemy* enemy) {
    std::cout << "Battle Start!" << std::endl;
    while (player->isAlive() && enemy->isAlive()) {
        randomEvent(player, enemy);
        int playerDamage = rand() % std::max(1, player->attack) + 1; // Ensure player's attack is at least 1
        int enemyDamage = rand() % std::max(1, enemy->attack) + 1; // Ensure enemy's attack is at least 1

        std::cout << player->name << " attacks " << enemy->name << " for " << playerDamage << " damage." << std::endl;
        enemy->takeDamage(playerDamage);
        std::cout << enemy->name << " attacks " << player->name << " for " << enemyDamage << " damage." << std::endl;
        player->takeDamage(enemyDamage);

        std::cout << player->status() << std::endl;
        std::cout << enemy->status() << std::endl;
        std::cout << std::endl;
    }

    if (!player->isAlive()) {
        std::cout << enemy->name << " wins!" << std::endl;
    }
    else if (!enemy->isAlive()) {
        std::cout << player->name << " wins!" << std::endl;
    }

    std::cout << "Game Over!" << std::endl;
}

int main() {
    srand(time(0));

    std::string playerName;
    std::cout << "Enter player name: ";
    std::cin >> playerName;
    Player player(playerName, 1);

    std::string enemyName = "Evil Wizard";
    Enemy enemy(enemyName, rand() % 3 + 3, "block");

    player.joinAlly(new Character("Knight", player.level));
    player.joinAlly(new Character("Archer", player.level));

    battle(&player, &enemy);

    std::cout << "Game Over!" << std::endl;
    std::cout << player.status() << std::endl;
    std::cout << enemy.status() << std::endl;

    return 0;
}

File: ChatGPT(예진)/Java/미니RPG전투게임.java
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

class Character {
    String name;
    int level;
    int maxHealth;
    int health;
    int attack;
    int experience;
    Map<String, Integer> items;
    Character[] allies;

    Character(String name, int level) {
        this.name = name;
        this.level = level;
        this.maxHealth = new Random().nextInt(51) + 50 + (10 * this.level);
        this.health = this.maxHealth;
        this.attack = new Random().nextInt(16) + 5 + (2 * this.level);
        this.experience = 0;
        this.items = new HashMap<>();
        this.items.put("health_potion", 3);
        this.items.put("mana_potion", 2);
        this.allies = new Character[]{};
    }

    void takeDamage(int damage) {
        this.health -= damage;
        if (this.health < 0) {
            this.health = 0;
        }
    }

    boolean isAlive() {
        return this.health > 0;
    }

    void levelUp() {
        this.level++;
        this.maxHealth += 10;
        this.attack += 2;
        this.health = this.maxHealth;
        System.out.println(this.name + " leveled up to level " + this.level + "!");
    }

    void gainExperience(int exp) {
        this.experience += exp;
        System.out.println(this.name + " gained " + exp + " experience points!");
        if (this.experience >= this.level * 100) {
            levelUp();
        }
    }

    void useItem(String item) {
        if (this.items.containsKey(item) && this.items.get(item) > 0) {
            if (item.equals("health_potion")) {
                int healthToRestore = Math.min(30, this.maxHealth - this.health);
                this.health += healthToRestore;
                this.items.put(item, this.items.get(item) - 1);
                System.out.println(this.name + " used a health potion and restored " + healthToRestore + " health!");
            } else if (item.equals("mana_potion")) {
                this.items.put(item, this.items.get(item) - 1);
                System.out.println(this.name + " used a mana potion!");
                // Add mana restoring logic here if needed
            }
        }
    }

    String status() {
        return this.name + "'s health: " + this.health + "/" + this.maxHealth + ", Level: " + this.level +
                ", Experience: " + this.experience + "/" + (this.level * 100) + ", Items: " + this.items;
    }

    void joinAlly(Character ally) {
        // Assuming that allies array will have fixed length
        Character[] temp = new Character[this.allies.length + 1];
        System.arraycopy(this.allies, 0, temp, 0, this.allies.length);
        temp[temp.length - 1] = ally;
        this.allies = temp;
        System.out.println(ally.name + " joined " + this.name + "'s party!");
    }
}

class Player extends Character {
    int gold;

    Player(String name, int level) {
        super(name, level);
        this.gold = 100;
    }

    @Override
    void takeDamage(int damage) {
        super.takeDamage(damage);
        if (this.attack < 0) {
            this.attack = 0;
        }
    }
}

class Enemy extends Character {
    String specialAbility;

    Enemy(String name, int level, String specialAbility) {
        super(name, level);
        this.specialAbility = specialAbility;
    }

    @Override
    void takeDamage(int damage) {
        if ("block".equals(this.specialAbility)) {
            System.out.println(this.name + " blocks the attack!");
        } else {
            super.takeDamage(damage);
        }
    }

    void attackTarget(Character target) {
        if ("stun".equals(this.specialAbility)) {
            System.out.println(this.name + " uses stun ability!");
            target.takeDamage(this.attack * 2);
        } else {
            target.takeDamage(Math.max(1, this.attack));
        }
    }
}

public class 미니RPG전투게임 {
    static Random random = new Random();

    static void randomEvent(Player player, Enemy enemy) {
        double eventChance = random.nextDouble(); // Generate a random number between 0 and 1
        if (eventChance < 0.2) { // 20% chance of an event occurring
            int event = random.nextInt(5) + 1;
            switch (event) {
                case 1:
                    System.out.println("A friendly wizard appears and heals you.");
                    player.health = Math.min(player.maxHealth, player.health + 20);
                    break;
                case 2:
                    System.out.println("You found a treasure chest! You gained 50 experience points.");
                    player.gainExperience(50);
                    break;
                case 3:
                    System.out.println("A curse befalls you, reducing your attack power!");
                    if (player.attack >= 5) {
                        player.attack -= 5;
                    } else {
                        player.attack = 0;
                    }
                    break;
                case 4:
                    System.out.println("You stumbled upon a hidden passage and found a health potion!");
                    player.useItem("health_potion");
                    break;
                default:
                    System.out.println("A mysterious mist surrounds you, confusing the enemy!");
                    if (enemy.attack >= 5) {
                        enemy.attack -= 5;
                    } else {
                        enemy.attack = 0;
                    }
                    break;
            }
        }
    }
    

    static void battle(Player player, Enemy enemy) {
        System.out.println("Battle Start!");
        while (player.isAlive() && enemy.isAlive()) {
            randomEvent(player, enemy);
            int playerDamage = random.nextInt(Math.max(1, player.attack)) + 1; // Ensure player's attack is at least 1
            int enemyDamage = random.nextInt(Math.max(1, enemy.attack)) + 1; // Ensure enemy's attack is at least 1
    
            System.out.println(player.name + " attacks " + enemy.name + " for " + playerDamage + " damage.");
            enemy.takeDamage(playerDamage);
            System.out.println(enemy.name + " attacks " + player.name + " for " + enemyDamage + " damage.");
            player.takeDamage(enemyDamage);
    
            System.out.println(player.status());
            System.out.println(enemy.status());
            System.out.println();
        }
    
        if (!player.isAlive()) {
            System.out.println(enemy.name + " wins!");
        } else if (!enemy.isAlive()) {
            System.out.println(player.name + " wins!");
        }
    
        System.out.println("Game Over!");
    }
    

    public static void main(String[] args) {
        java.util.Scanner scanner = new java.util.Scanner(System.in);
        System.out.print("Enter player name: ");
        String playerName = scanner.nextLine();
        Player player = new Player(playerName, 1);

        String enemyName = "Evil Wizard";
        Enemy enemy = new Enemy(enemyName, random.nextInt(3) + 3, "block");

        player.joinAlly(new Character("Knight", player.level));
        player.joinAlly(new Character("Archer", player.level));

        battle(player, enemy);

        System.out.println("Game Over!");
        System.out.println(player.status());
        System.out.println(enemy.status());

        scanner.close();
    }
}
